File: leetcode/src/00137_single_number_ii/solution.go
package leetcode

func singleNumber(nums []int) int {
	res := 0
	for i := 0; i < 64; i++ {
		cnt := 0
		mask := 1 << i
		for _, num := range nums {
			if num&mask != 0 {
				cnt++
			}
		}
		if cnt > 0 && cnt%3 != 0 {
			res |= mask
		}
	}
	return res
}

File: leetcode/src/00137_single_number_ii/solution.rs
struct Solution;

impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        let (mut c1, mut c2) = (0, 0);
        for &num in &nums {
            c2 ^= c1 & num;
            c1 ^= num;
            let mask = !(c1 & c2);
            c1 &= mask;
            c2 &= mask;
        }
        c1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        let nums = vec![2, 2, 3, 2];
        assert_eq!(Solution::single_number(nums), 3);

        let nums = vec![0, 1, 0, 1, 0, 1, 99];
        assert_eq!(Solution::single_number(nums), 99);
    }
}
